### PR TITLE
[Doc] buildpack ref docs

### PR DESCRIPTION
**Pull request by jcalcaben**
## Description

This PR updates the JSDocs in the targets in the Buildpack package.
It also creates a file in devdocs that pulls this information into a reference-style topic.

## Related Issue

Closes PWA-703

## Acceptance 

@zetlen 

### Verification Stakeholders

@zetlen 

### Verification Steps

1. Navigate to the pwa devdocs directory: `cd pwa-devdocs`
1. Start the HTML preview server: `yarn develop`
1. Verify files are auto-generated:

    - `src/_includes/auto-generated/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.md`
    - `src/_includes/auto-generated/pwa-buildpack/lib/BuildBus/declare-base.md`
    - `src/_includes/auto-generated/pwa-buildpack/lib/BuildBus/BuildBus.md`
    - `src/_includes/auto-generated/pwa-buildpack/lib/BuildBus/TargetProvider.md`
    - `src/_includes/auto-generated/pwa-buildpack/lib/BuildBus/Target.md`
    - `src/_includes/auto-generated/pwa-buildpack/lib/Utilities/getEnvVarDefinitions.js`

1. On the devdocs preview site, in the top navigation, click on **PWA Buildpack** under the _Reference API_ item.
1. Verify new navigation layout
1. Verify all the links under the Extension points, Extension framework, Developer utilities, Webpack tools, and Environment configuration sections work.

## Checklist

- [ ] I have added tests to cover my changes, if necessary.
* I have updated the documentation accordingly, if necessary.

### CODE REVIEW

**Comment by devops-pwa-codebuild** [comment]
**Performance Test Results**

The following fails have been reported by WebpageTest. These numbers indicates a possible performance issue with the PR which requires further manual testing to validate.

https://pr-2508.pwa-venia.com : LH Performance Expected 0.85 Actual 0.57, LH Best Practices Expected 1 Actual 0.92
https://pr-2508.pwa-venia.com/venia-tops.html : LH Performance Expected 0.75 Actual 0.35, LH Best Practices Expected 1 Actual 0.92
https://pr-2508.pwa-venia.com/valeria-two-layer-tank.html : LH Performance Expected 0.8 Actual 0.49, LH Accessibility Expected 0.9 Actual 0.89, LH Best Practices Expected 1 Actual 0.92

File: packages/pwa-buildpack/lib/BuildBus/Target.js [hunk 1]
```diff
@@ -124,17 +136,43 @@ class Target extends Trackable {
         });
     }
     /**
-     *
+     *  Runs the `tap()` function on a target
+     * 
+     * @param {string|object} name string or object containing the name of the interceptor
```
[the remark this answers is not in the document]
Comment (jcalcaben) [reply packages/pwa-buildpack/lib/BuildBus/Target.js:141]: we can add that information to the description

File: packages/pwa-buildpack/lib/BuildBus/declare-base.js [hunk 2]
```diff
@@ -52,30 +48,23 @@ module.exports = targets => {
...
          * Collects requests to intercept and modify individual files from this
-         * dependency. Only files from the currently requesting dependency may
-         * be transformed.
-         * **This is a very low-level extension point; it should be used as a
+         * dependency.
+         *
+         * Since the storefront developer is in charge of important dependencies, 
+         * the interceptor files in the storefront project itself should be able to
+         * transform ANY file from ANY dependency.
+         * However, interceptor files in the storefront dependencies are prevented
```
Comment (zetlen) [inline packages/pwa-buildpack/lib/BuildBus/declare-base.js:58]: 👍

File: packages/pwa-buildpack/lib/BuildBus/mapHooksToTargets.js [hunk 3]
```diff
@@ -33,6 +33,7 @@ const VALID_TYPES = new Map();
  * Dictionary of Tapable Hook classes to expose under these new names.
  * @public
  * @type {Object.<string,Tapable.Hook>}
+ * @see [Tapable]{@link https://github.com/webpack/tapable}
```
Comment (zetlen) [inline packages/pwa-buildpack/lib/BuildBus/mapHooksToTargets.js:36]: You know what? I don't think this file or these functions should be public. They're not organized along with the rest of the Buildpack API, so they `mapHooksToTargets` looks weird in the navigation.

I shouldn't have set the @public keywords on these functions. They're utility functions that BuildBus uses. Sorry about that.

File: pwa-devdocs/scripts/create-reference-docs/config/buildpack.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        target: 'pwa-buildpack/lib/BuildBus/BuildBus.js',
+        type: 'function'
+    },
+    {
+        target: 'pwa-buildpack/lib/BuildBus/Target.js',
+        type: 'function'
+    },
+    {
+        target: 'pwa-buildpack/lib/BuildBus/mapHooksToTargets.js',
+        type: 'function'
```
Comment (zetlen) [inline pwa-devdocs/scripts/create-reference-docs/config/buildpack.js:24]:
```suggestion
```

File: pwa-devdocs/scripts/create-reference-docs/config/buildpack.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+module.exports = [
+    {
+        target: 'pwa-buildpack/lib/BuildBus/declare-base.js',
+        type: 'function'
+    },
```
Comment (zetlen) [inline pwa-devdocs/scripts/create-reference-docs/config/buildpack.js:5]: Does this belong here? The `declare-base.js` file is what declares Buildpack targets, so we wrote the doc blocks in that file. But the targets themselves are in a different category of thing, which should be documented in a separate "Targets" section, right?

File: pwa-devdocs/src/_data/buildpack-api.yml [hunk 6]
```diff
@@ -1,5 +1,17 @@
 title: PWA Buildpack API
 entries:
+    - label: BuildBus
+      url: /pwa-buildpack/reference/buildbus/
+      entries:
+        - label: Extensibility Targets
+          url: /pwa-buildpack/reference/buildbus/targets/
+        - label: mapHooksToTarget
+          url: /pwa-buildpack/reference/buildbus/maphookstotargets/
```
Comment (zetlen) [inline pwa-devdocs/src/_data/buildpack-api.yml:9]:
```suggestion
```

File: pwa-devdocs/src/_data/buildpack-api.yml [hunk 7]
```diff
@@ -1,5 +1,17 @@
 title: PWA Buildpack API
 entries:
+    - label: BuildBus
+      url: /pwa-buildpack/reference/buildbus/
+      entries:
+        - label: Extensibility Targets
+          url: /pwa-buildpack/reference/buildbus/targets/
```
Comment (zetlen) [inline pwa-devdocs/src/_data/buildpack-api.yml:7]: I know that I called the directory `BuildBus` in source code, but I think in documentation, the top-level item should be called "Extensibility" and there should be an additional entry below, called "BuildBus".

File: pwa-devdocs/src/_data/buildpack-api.yml [hunk 8]
```diff
@@ -24,6 +36,9 @@ entries:
     - label: MagentoResolver
       url: /pwa-buildpack/reference/magento-resolver/
 
+    - label: ModuleTransformConfig
+      url: /pwa-buildpack/reference/moduletransformconfig/
+
```
Comment (zetlen) [inline pwa-devdocs/src/_data/buildpack-api.yml:41]: This belongs under Extensibility

File: pwa-devdocs/scripts/create-reference-docs/config/buildpack.js (hunk 5 above)
Comment (jcalcaben) [reply pwa-devdocs/scripts/create-reference-docs/config/buildpack.js:5]: I can't add another level of nesting to the navigation, so they will all need to go under "Extensibility"

File: packages/pwa-buildpack/lib/BuildBus/Target.js [hunk 9]
```diff
@@ -17,7 +17,6 @@ const interceptionTypes = {
  * Target instead of the literal Tapable instance. This enables
  * better logging, error checking, and validation.
  *
- * @class Target
  * @extends {Trackable}
  */
 class Target extends Trackable {
```
[the remark this answers is not in the document]
Comment (jcalcaben) [reply packages/pwa-buildpack/lib/BuildBus/Target.js:24]: Link added

File: packages/pwa-buildpack/lib/BuildBus/mapHooksToTargets.js (hunk 3 above)
Comment (jcalcaben) [reply packages/pwa-buildpack/lib/BuildBus/mapHooksToTargets.js:36]: topic file deleted

**Comment by dpatil-magento** [comment]
Verification steps look good.

**Pull request merged**